lib/client.js

```diff
@@ -83,6 +85,14 @@ Client.prototype._connect = function (callback) {
   }
   this._connecting = true
 
+  var connectionTimeoutHandle
+  if (this._connectionTimeoutMillis > 0) {
+    connectionTimeoutHandle = setTimeout(() => {
+      con._ending = true
+      con.stream.destroy(new Error('timeout expired'))
+    }, this._connectionTimeoutMillis)
+  }
+
```

Comment:
I think that `con.once('end', () => {` also should have `clearTimeout(connectionTimeoutHandle)`, use case:
`connectionTimeoutMillis` set to some value.
Application receive SIGTERM or any other shutdown signal, because client still not connected we can not call `client.end()`, so application can call `client.connection.stream.destroy()`.
If `destroy` will be called with error, then `error` event will be raised, `connectingErrorHandler` will be executed and timeout `connectionTimeoutHandle` will be cleared. Everything OK.
If `destroy` will be called without arguments, `clearTimeout(connectionTimeoutHandle)` will never happened and applucation will be hang up while timeout will not be called.

---
Why we can not call `client.end()`:
Current end code:
```js
  if (this.activeQuery) {
    // if we have an active query we need to force a disconnect
    // on the socket - otherwise a hung query could block end forever
    this.connection.stream.destroy()
  } else {
    this.connection.end()
  }
```
Because `activeQuery` is undefined while client is not connected, `connection.end()` will be called, which itself implies opened connection, because write data to socket.

Reply:
You are correct; the timer should be cleared in that event handler too.

To make sure I understand your use case, are you calling `client.connection.stream.destroy()` in your signal handler if the connection hasn't been established yet, or is there another way to trigger this behavior?
And passing in an error, e.g. `client.connection.stream.destroy(new Error('SIGTERM'))` is the workaround?

Reply:
Currently I do not call `stream.destroy()`, but I'd like modify application, because hang up is not cool.
`stream.destroy(new Error(...))` is workaround, yes. But from other side, destroy called from `client.end` called without arguments... (which will not work actually in current case, just as example).

Reply:
I did some testing just to confirm.

Calling `client.end()` before the connection is established seems to wait for the connection to succeed (in which case the timeout will be cleared). There will be no `activeQuery` before connecting successfully.

Calling `client.connection.stream.destroy()` (without an error) will reject the `client.connect()` promise with a `Connection terminated unexpectedly` error and wait for the timeout.

Calling `client.connection.stream.destroy(new Error())` (with an error) will reject the `client.connect()` promise with the passed in error and not wait for the timeout.

The only way to trigger this error is to call `client.connection.stream.destroy()` from your own code. And the solution is to pass in an error.